src/extra/gpkg/read.jl

```diff
@@ -0,0 +1,447 @@
+    end
+  end
+  
+  # Query and process remaining rows
+  table_result = DBInterface.execute(db, "SELECT * FROM \"$table_name\"")
+  row_count = 0
+  for row in Tables.rows(table_result)
+    row_count += 1
+    if row_count == 1
+      continue  # Skip first row since we already processed it
+    end
+    
+    for col_name in all_columns
+      col_symbol = Symbol(col_name)
+      if col_name == geom_column
+        # Parse geometry blob
+        blob = Tables.getcolumn(row, col_symbol)
+        if ismissing(blob) || isnothing(blob)
+          push!(column_data[col_symbol], blob)
+        else
+          push!(column_data[col_symbol], parse_gpb(blob, crs_type))
+        end
+      else
+        # Store attribute data as-is
+        push!(column_data[col_symbol], Tables.getcolumn(row, col_symbol))
+      end
+    end
+  end
+  
+  # 4. Build final table
+  final_names = Symbol[]
+  final_columns = Any[]
+  
+  for col_name in all_columns
+    if col_name == geom_column
+      push!(final_names, :geometry)
+    else
+      push!(final_names, Symbol(col_name))
+    end
+    push!(final_columns, column_data[Symbol(col_name)])
+  end
+  
+  close(db)
+  
+  return (; zip(final_names, final_columns)...)
+end
+
+function gpkgmetadata(db::SQLite.DB, layer::Int)
+  # Get feature table names
+  contents_result = DBInterface.execute(db,
+    """
+    SELECT table_name, data_type
+    FROM gpkg_contents
+    WHERE data_type = 'features'
+    ORDER BY table_name
+    """)
+  
+  # Use Iterators.drop to get the requested layer
+  rows = Tables.rows(contents_result)
+  layer_iter = Iterators.drop(rows, layer - 1)
+  
+  if isempty(layer_iter)
+    error("Layer $layer not found in GeoPackage")
+  end
+  first_row = first(layer_iter)
+  
+  table_name = first_row.table_name
+  
+  # Get geometry column information
+  geom_result = DBInterface.execute(db,
+    """
+    SELECT column_name, srs_id
+    FROM gpkg_geometry_columns
+    WHERE table_name = ?
+    """, [table_name])
+  # Get geometry column information using iterate
+  geom_state = iterate(Tables.rows(geom_result))
+  if isnothing(geom_state)
+    error("No geometry column found for table $table_name")
+  end
+  geom_info = geom_state[1]
+  geom_column = geom_info.column_name
+  srs_id = geom_info.srs_id
+  crs_type = get_crs_from_srid(db, srs_id, table_name, geom_column)
+  
+  return table_name, geom_column, crs_type
+end
+
+
+# Get CRS from SRID according to GeoPackage specification
+# Reference: https://www.geopackage.org/spec/#gpkg_spatial_ref_sys_cols
+# SRID possibilities:
+#  0: Undefined geographic coordinate reference system
+# -1: Undefined Cartesian coordinate reference system  
+#  1-32766: Reserved for OGC use (EPSG codes)
+#  32767-65535: Reserved for GeoPackage use
+#  >65535: User-defined coordinate reference systems
+function get_crs_from_srid(db::SQLite.DB, srid::Integer, table_name::String, geom_column::String)
```

Comment:
I believe the encapsulation of concepts can be improved. The body of this function should probably be part of the body of the gpkgmetadata function.